File: OtherLLMEngines/Glm4.7/board.py
# board.py

import copy

# Константы для фигур
EMPTY = 0
wP, wN, wB, wR, wQ, wK = 1, 2, 3, 4, 5, 6
bP, bN, bB, bR, bQ, bK = 9, 10, 11, 12, 13, 14

# Словари для преобразования
PIECE_SYMBOLS = {
    1: 'P', 2: 'N', 3: 'B', 4: 'R', 5: 'Q', 6: 'K',
    9: 'p', 10: 'n', 11: 'b', 12: 'r', 13: 'q', 14: 'k'
}
COLORS = {wP: 'w', wN: 'w', wB: 'w', wR: 'w', wQ: 'w', wK: 'w',
          bP: 'b', bN: 'b', bB: 'b', bR: 'b', bQ: 'b', bK: 'b'}

class Board:
    """Класс для представления и управления шахматной доской."""
    def __init__(self, fen=None):
        if fen:
            self.load_fen(fen)
        else:
            self.reset_to_start_pos()

    def reset_to_start_pos(self):
        """Устанавливает начальную позицию."""
        self.load_fen("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1")

    def load_fen(self, fen_str):
        """Загружает позицию из строки FEN."""
        parts = fen_str.split()
        # Расстановка фигур
        piece_placement = parts[0]
        self.squares = [EMPTY] * 64
        rank = 7
        file = 0
        for char in piece_placement:
            if char == '/':
                rank -= 1
                file = 0
            elif char.isdigit():
                file += int(char)
            else:
                piece_char = char.lower()
                color = 'w' if char.isupper() else 'b'
                piece_map = {'p': wP, 'n': wN, 'b': wB, 'r': wR, 'q': wQ, 'k': wK}
                piece = piece_map[piece_char]
                if color == 'b':
                    piece += 8
                self.squares[rank * 8 + file] = piece
                file += 1
        
        # Ход
        self.turn = parts[1]
        
        # Правила рокировки
        self.castle_rights = parts[2]
        
        # Взятие на проходе
        self.en_passant_sq = -1 if parts[3] == '-' else self.algebraic_to_index(parts[3])
        
        # Счетчик полуходов (правило 50 ходов)
        self.halfmove_clock = int(parts[4])
        
        # Номер хода
        self.fullmove_number = int(parts[5])

        # История позиций для определения ничьи по троекратному повторению
        self.position_history = [self.get_board_hash()]

    def to_fen(self):
        """Преобразует текущую позицию в строку FEN."""
        # Расстановка
        fen_parts = []
        for rank in range(8):
            empty_count = 0
            rank_str = ""
            for file in range(8):
                piece = self.squares[rank * 8 + file]
                if piece == EMPTY:
                    empty_count += 1
                else:
                    if empty_count > 0:
                        rank_str += str(empty_count)
                        empty_count = 0
                    rank_str += PIECE_SYMBOLS[piece]
            if empty_count > 0:
                rank_str += str(empty_count)
            fen_parts.append(rank_str)
        fen = "/".join(fen_parts)
        
        # Ход
        fen += f" {self.turn}"
        
        # Рокировка
        fen += f" {self.castle_rights}"
        
        # Взятие на проходе
        ep_str = self.index_to_algebraic(self.en_passant_sq) if self.en_passant_sq != -1 else '-'
        fen += f" {ep_str}"
        
        # Счетчики
        fen += f" {self.halfmove_clock} {self.fullmove_number}"
        return fen

    def algebraic_to_index(self, alg):
        """Конвертирует алгебраическую нотацию (e.g., 'e2') в индекс (0-63)."""
        file = ord(alg[0]) - ord('a')
        rank = int(alg[1]) - 1
        return rank * 8 + file

    def index_to_algebraic(self, idx):
        """Конвертирует индекс (0-63) в алгебраическую нотацию."""
        if idx == -1: return '-'
        rank = idx // 8
        file = idx % 8
        return chr(ord('a') + file) + str(rank + 1)

    def get_piece(self, sq):
        """Возвращает фигуру на клетке."""
        return self.squares[sq]

    def set_piece(self, sq, piece):
        """Устанавливает фигуру на клетку."""
        self.squares[sq] = piece

    def make_move(self, move):
        """Выполняет ход на доске."""
        from_sq, to_sq, promotion_piece = move
        piece = self.get_piece(from_sq)
        captured_piece = self.get_piece(to_sq)
        
        # Сохраняем состояние для undo
        undo_info = {
            'move': move,
            'captured_piece': captured_piece,
            'castle_rights': self.castle_rights,
            'en_passant_sq': self.en_passant_sq,
            'halfmove_clock': self.halfmove_clock,
            'fullmove_number': self.fullmove_number,
        }

        # Обновляем счетчик полуходов
        if (piece in (wP, bP)) or (captured_piece != EMPTY):
            self.halfmove_clock = 0
        else:
            self.halfmove_clock += 1
        
        # Обновляем номер полного хода
        if self.turn == 'b':
            self.fullmove_number += 1
        
        # Сброс взятия на проходе
        self.en_passant_sq = -1

        # Специальные ходы
        if piece in (wP, bP):
            # Ход на 2 клетки -> устанавливаем возможность en passant
            if abs(to_sq - from_sq) == 16:
                self.en_passant_sq = (from_sq + to_sq) // 2
            # Взятие на проходе
            elif to_sq == self.en_passant_sq:
                captured_pawn_sq = to_sq + 8 if piece == wP else to_sq - 8
                self.set_piece(captured_pawn_sq, EMPTY)
            # Превращение
            elif to_sq // 8 == 0 or to_sq // 8 == 7:
                piece = promotion_piece

        # Рокировка
        if piece in (wK, bK) and abs(to_sq - from_sq) == 2:
            is_kingside = to_sq > from_sq
            rook_from = to_sq + 1 if is_kingside else to_sq - 2
            rook_to = to_sq - 1 if is_kingside else to_sq + 1
            self.set_piece(rook_to, self.get_piece(rook_from))
            self.set_piece(rook_from, EMPTY)
        
        # Обновление прав рокировки
        if piece == wK: self.castle_rights = self.castle_rights.replace('K', '').replace('Q', '')
        if piece == bK: self.castle_rights = self.castle_rights.replace('k', '').replace('q', '')
        if piece == wR:
            if from_sq == 63: self.castle_rights = self.castle_rights.replace('K', '')
            if from_sq == 56: self.castle_rights = self.castle_rights.replace('Q', '')
        if piece == bR:
            if from_sq == 7: self.castle_rights = self.castle_rights.replace('k', '')
            if from_sq == 0: self.castle_rights = self.castle_rights.replace('q', '')

        # Выполнение основного хода
        self.set_piece(to_sq, piece)
        self.set_piece(from_sq, EMPTY)
        
        # Смена хода
        self.turn = 'b' if self.turn == 'w' else 'w'
        
        # Добавляем хэш позиции в историю
        self.position_history.append(self.get_board_hash())
        
        return undo_info

    def unmake_move(self, undo_info):
        """Отменяет ход, используя сохраненную информацию."""
        move = undo_info['move']
        from_sq, to_sq, promotion_piece = move
        
        # Восстанавливаем счетчики и флаги
        self.castle_rights = undo_info['castle_rights']
        self.en_passant_sq = undo_info['en_passant_sq']
        self.halfmove_clock = undo_info['halfmove_clock']
        self.fullmove_number = undo_info['fullmove_number']
        
        # Возвращаем ход
        self.turn = 'b' if self.turn == 'w' else 'w'
        
        piece = self.get_piece(to_sq)
        captured_piece = undo_info['captured_piece']
        
        # Отмена рокировки
        if piece in (wK, bK) and abs(to_sq - from_sq) == 2:
            is_kingside = to_sq > from_sq
            rook_from = to_sq + 1 if is_kingside else to_sq - 2
            rook_to = to_sq - 1 if is_kingside else to_sq + 1
            self.set_piece(rook_from, self.get_piece(rook_to))
            self.set_piece(rook_to, EMPTY)

        # Отмена взятия на проходе
        if piece in (wP, bP) and to_sq == undo_info['en_passant_sq']:
            captured_pawn_sq = to_sq + 8 if piece == wP else to_sq - 8
            self.set_piece(captured_pawn_sq, bP if piece == wP else wP)
            self.set_piece(to_sq, EMPTY)
        else:
            # Отмена превращения пешки
            if promotion_piece:
                piece = wP if self.turn == 'w' else bP
            
            # Отмена обычного хода/взятия
            self.set_piece(from_sq, piece)
            self.set_piece(to_sq, captured_piece)
        
        # Удаляем последний хэш из истории
        self.position_history.pop()

    def is_square_attacked(self, sq, by_color):
        """Проверяет, атакована ли клетка sq фигурами цвета by_color."""
        # Пешки
        pawn_dir = -1 if by_color == 'w' else 1
        if 0 <= sq + pawn_dir * 8 - 1 < 64 and (sq % 8) != 0:
            if self.get_piece(sq + pawn_dir * 8 - 1) == (wP if by_color == 'w' else bP):
                return True
        if 0 <= sq + pawn_dir * 8 + 1 < 64 and (sq % 8) != 7:
            if self.get_piece(sq + pawn_dir * 8 + 1) == (wP if by_color == 'w' else bP):
                return True

        # Конь
        knight_offsets = [-17, -15, -10, -6, 6, 10, 15, 17]
        for offset in knight_offsets:
            target_sq = sq + offset
            if 0 <= target_sq < 64 and abs(sq % 8 - target_sq % 8) <= 2:
                if self.get_piece(target_sq) == (wN if by_color == 'w' else bN):
                    return True
        
        # Слоники, Ладьи, Ферзи (скользящие фигуры)
        sliding_offsets = {
            (wB, bR): [(-1, -1), (-1, 1), (1, -1), (1, 1)], # Слон
            (wR, bB): [(-1, 0), (1, 0), (0, -1), (0, 1)],   # Ладья
            (wQ, bQ): [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)] # Ферзь
        }
        for pieces, offsets in sliding_offsets.items():
            for dr, dc in offsets:
                for i in range(1, 8):
                    target_sq = sq + dr * 8 + dc
                    if not (0 <= target_sq < 64): break
                    if abs((sq % 8) - (target_sq % 8)) != abs(dc) * i and dc != 0: break
                    
                    piece_on_target = self.get_piece(target_sq)
                    if piece_on_target == EMPTY: continue
                    
                    if (by_color == 'w' and piece_on_target in pieces[0]) or \
                       (by_color == 'b' and piece_on_target in pieces[1]):
                        return True
                    break
        
        # Король
        king_offsets = [-9, -8, -7, -1, 1, 7, 8, 9]
        for offset in king_offsets:
            target_sq = sq + offset
            if 0 <= target_sq < 64 and abs(sq % 8 - target_sq % 8) <= 1:
                if self.get_piece(target_sq) == (wK if by_color == 'w' else bK):
                    return True
        
        return False

    def get_board_hash(self):
        """Возвращает хэш текущей позиции (используется для определения повторения)."""
        # Для простоты используем FEN как хэш. В реальных движках используются Zobrist keys.
        return f"{self.to_fen()}_{self.turn}"

File: rust_engine/src/board.rs
//! OpusChess - Board Representation Module
//!
//! This module provides the core data structures for representing a chess board,
//! pieces, and moves. It includes FEN parsing and generation, move execution,
//! and position history tracking.

use crate::types::*;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

/// Starting position FEN
pub const STARTING_FEN: &str = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1";

/// Represents a chess move
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct Move {
    pub from_sq: usize,
    pub to_sq: usize,
    pub promotion: u8,
    pub is_castling: bool,
    pub is_en_passant: bool,
}

impl Move {
    /// Create a new move
    pub fn new(from_sq: usize, to_sq: usize) -> Self {
        Move {
            from_sq,
            to_sq,
            promotion: 0,
            is_castling: false,
            is_en_passant: false,
        }
    }

    /// Create a promotion move
    pub fn with_promotion(from_sq: usize, to_sq: usize, promotion: u8) -> Self {
        Move {
            from_sq,
            to_sq,
            promotion,
            is_castling: false,
            is_en_passant: false,
        }
    }

    /// Create a castling move
    pub fn castling(from_sq: usize, to_sq: usize) -> Self {
        Move {
            from_sq,
            to_sq,
            promotion: 0,
            is_castling: true,
            is_en_passant: false,
        }
    }

    /// Create an en passant move
    pub fn en_passant(from_sq: usize, to_sq: usize) -> Self {
        Move {
            from_sq,
            to_sq,
            promotion: 0,
            is_castling: false,
            is_en_passant: true,
        }
    }

    /// Convert move to UCI notation (e.g., "e2e4", "e7e8q")
    pub fn to_uci(&self) -> String {
        let mut uci = format!("{}{}", square_name(self.from_sq), square_name(self.to_sq));
        if self.promotion != 0 {
            let promo_char = match self.promotion {
                QUEEN => 'q',
                ROOK => 'r',
                BISHOP => 'b',
                KNIGHT => 'n',
                _ => return uci,
            };
            uci.push(promo_char);
        }
        uci
    }

    /// Null move constant
    pub fn null() -> Self {
        Move::new(0, 0)
    }

    /// Check if this is a null move
    pub fn is_null(&self) -> bool {
        self.from_sq == 0 && self.to_sq == 0 && self.promotion == 0
    }
}

impl Default for Move {
    fn default() -> Self {
        Move::null()
    }
}

/// Information needed to undo a move
#[derive(Clone, Copy, Debug)]
pub struct UndoInfo {
    pub captured_piece: u8,
    pub castling_rights: u8,
    pub en_passant_square: i8,
    pub halfmove_clock: u16,
    pub moved_piece: u8,
}

/// Chess board representation
#[derive(Clone)]
pub struct Board {
    /// 64-element array representing the board (0=a1, 1=b1, ..., 63=h8)
    pub squares: [u8; 64],
    /// True if it's white's turn
    pub white_to_move: bool,
    /// Bitmask for castling rights (1=K, 2=Q, 4=k, 8=q)
    pub castling_rights: u8,
    /// Target square for en passant (-1 if none)
    pub en_passant_square: i8,
    /// Moves since last pawn move or capture (for 50-move rule)
    pub halfmove_clock: u16,
    /// Full move counter
    pub fullmove_number: u16,
    /// Position history for repetition detection
    pub position_history: Vec<u64>,
}

impl Board {
    /// Create a new board with the starting position
    pub fn new() -> Self {
        Board::from_fen(STARTING_FEN).unwrap()
    }

    /// Create a board from a FEN string
    pub fn from_fen(fen: &str) -> Option<Self> {
        let parts: Vec<&str> = fen.split_whitespace().collect();
        if parts.is_empty() {
            return None;
        }

        let mut board = Board {
            squares: [EMPTY; 64],
            white_to_move: true,
            castling_rights: 0,
            en_passant_square: -1,
            halfmove_clock: 0,
            fullmove_number: 1,
            position_history: Vec::new(),
        };

        // Parse piece placement
        let mut rank = 7i8;
        let mut file = 0i8;
        
        for c in parts[0].chars() {
            if c == '/' {
                rank -= 1;
                file = 0;
            } else if c.is_ascii_digit() {
                file += c.to_digit(10).unwrap() as i8;
            } else if let Some(piece) = fen_to_piece(c) {
                let sq = (rank * 8 + file) as usize;
                if sq < 64 {
                    board.squares[sq] = piece;
                }
                file += 1;
            }
        }

        // Parse active color
        if parts.len() > 1 {
            board.white_to_move = parts[1] != "b";
        }

        // Parse castling rights
        if parts.len() > 2 && parts[2] != "-" {
            for c in parts[2].chars() {
                match c {
                    'K' => board.castling_rights |= CASTLE_WK,
                    'Q' => board.castling_rights |= CASTLE_WQ,
                    'k' => board.castling_rights |= CASTLE_BK,
                    'q' => board.castling_rights |= CASTLE_BQ,
                    _ => {}
                }
            }
        }

        // Parse en passant square
        if parts.len() > 3 && parts[3] != "-" {
            if let Some(sq) = parse_square(parts[3]) {
                board.en_passant_square = sq as i8;
            }
        }

        // Parse halfmove clock
        if parts.len() > 4 {
            board.halfmove_clock = parts[4].parse().unwrap_or(0);
        }

        // Parse fullmove number
        if parts.len() > 5 {
            board.fullmove_number = parts[5].parse().unwrap_or(1);
        }

        // Initialize position history
        board.position_history.push(board.compute_hash());

        Some(board)
    }

    /// Generate FEN string from current board state
    pub fn to_fen(&self) -> String {
        let mut fen = String::new();

        // Piece placement
        for rank in (0..8).rev() {
            let mut empty_count = 0;
            for file in 0..8 {
                let piece = self.squares[rank * 8 + file];
                if piece == EMPTY {
                    empty_count += 1;
                } else {
                    if empty_count > 0 {
                        fen.push_str(&empty_count.to_string());
                        empty_count = 0;
                    }
                    if let Some(c) = piece_to_fen(piece) {
                        fen.push(c);
                    }
                }
            }
            if empty_count > 0 {
                fen.push_str(&empty_count.to_string());
            }
            if rank > 0 {
                fen.push('/');
            }
        }

        // Active color
        fen.push(' ');
        fen.push(if self.white_to_move { 'w' } else { 'b' });

        // Castling rights
        fen.push(' ');
        if self.castling_rights == 0 {
            fen.push('-');
        } else {
            if self.castling_rights & CASTLE_WK != 0 { fen.push('K'); }
            if self.castling_rights & CASTLE_WQ != 0 { fen.push('Q'); }
            if self.castling_rights & CASTLE_BK != 0 { fen.push('k'); }
            if self.castling_rights & CASTLE_BQ != 0 { fen.push('q'); }
        }

        // En passant
        fen.push(' ');
        if self.en_passant_square >= 0 {
            fen.push_str(&square_name(self.en_passant_square as usize));
        } else {
            fen.push('-');
        }

        // Halfmove clock and fullmove number
        fen.push_str(&format!(" {} {}", self.halfmove_clock, self.fullmove_number));

        fen
    }

    /// Compute a hash of the current position for repetition detection
    fn compute_hash(&self) -> u64 {
        let mut hasher = DefaultHasher::new();
        self.squares.hash(&mut hasher);
        self.white_to_move.hash(&mut hasher);
        self.castling_rights.hash(&mut hasher);
        self.en_passant_square.hash(&mut hasher);
        hasher.finish()
    }

    /// Execute a move on the board. Returns UndoInfo for undoing the move later.
    pub fn make_move(&mut self, mv: &Move) -> UndoInfo {
        let from_sq = mv.from_sq;
        let to_sq = mv.to_sq;
        let piece = self.squares[from_sq];
        let captured = self.squares[to_sq];

        // Save undo information
        let undo = UndoInfo {
            captured_piece: if mv.is_en_passant {
                if self.white_to_move { BLACK_PAWN } else { WHITE_PAWN }
            } else {
                captured
            },
            castling_rights: self.castling_rights,
            en_passant_square: self.en_passant_square,
            halfmove_clock: self.halfmove_clock,
            moved_piece: piece,
        };

        // Update halfmove clock
        let piece_type = get_piece_type(piece);
        if piece_type == PAWN || captured != EMPTY {
            self.halfmove_clock = 0;
        } else {
            self.halfmove_clock += 1;
        }

        // Handle en passant capture
        if mv.is_en_passant {
            if self.white_to_move {
                self.squares[to_sq - 8] = EMPTY;
            } else {
                self.squares[to_sq + 8] = EMPTY;
            }
        }

        // Handle castling
        if mv.is_castling {
            match to_sq {
                6 => {  // White kingside (g1)
                    self.squares[7] = EMPTY;
                    self.squares[5] = WHITE_ROOK;
                }
                2 => {  // White queenside (c1)
                    self.squares[0] = EMPTY;
                    self.squares[3] = WHITE_ROOK;
                }
                62 => { // Black kingside (g8)
                    self.squares[63] = EMPTY;
                    self.squares[61] = BLACK_ROOK;
                }
                58 => { // Black queenside (c8)
                    self.squares[56] = EMPTY;
                    self.squares[59] = BLACK_ROOK;
                }
                _ => {}
            }
        }

        // Move the piece
        self.squares[to_sq] = piece;
        self.squares[from_sq] = EMPTY;

        // Handle promotion
        if mv.promotion != 0 {
            self.squares[to_sq] = (if self.white_to_move { WHITE } else { BLACK }) | mv.promotion;
        }

        // Update castling rights
        if piece_type == KING {
            if self.white_to_move {
                self.castling_rights &= !(CASTLE_WK | CASTLE_WQ);
            } else {
                self.castling_rights &= !(CASTLE_BK | CASTLE_BQ);
            }
        }

        // If rook moves or is captured
        if from_sq == 0 || to_sq == 0 { self.castling_rights &= !CASTLE_WQ; }
        if from_sq == 7 || to_sq == 7 { self.castling_rights &= !CASTLE_WK; }
        if from_sq == 56 || to_sq == 56 { self.castling_rights &= !CASTLE_BQ; }
        if from_sq == 63 || to_sq == 63 { self.castling_rights &= !CASTLE_BK; }

        // Update en passant square
        self.en_passant_square = -1;
        if piece_type == PAWN {
            let diff = (to_sq as i32) - (from_sq as i32);
            if diff.abs() == 16 {
                self.en_passant_square = ((from_sq as i32 + to_sq as i32) / 2) as i8;
            }
        }

        // Update fullmove number
        if !self.white_to_move {
            self.fullmove_number += 1;
        }

        // Switch side to move
        self.white_to_move = !self.white_to_move;

        // Update position history
        self.position_history.push(self.compute_hash());

        undo
    }

    /// Undo a move using saved UndoInfo
    pub fn unmake_move(&mut self, mv: &Move, undo: &UndoInfo) {
        // Switch side back
        self.white_to_move = !self.white_to_move;

        let from_sq = mv.from_sq;
        let to_sq = mv.to_sq;

        // Restore the moved piece
        self.squares[from_sq] = undo.moved_piece;

        // Restore captured piece
        if mv.is_en_passant {
            self.squares[to_sq] = EMPTY;
            if self.white_to_move {
                self.squares[to_sq - 8] = BLACK_PAWN;
            } else {
                self.squares[to_sq + 8] = WHITE_PAWN;
            }
        } else {
            self.squares[to_sq] = undo.captured_piece;
        }

        // Handle castling - move rook back
        if mv.is_castling {
            match to_sq {
                6 => {  // White kingside
                    self.squares[5] = EMPTY;
                    self.squares[7] = WHITE_ROOK;
                }
                2 => {  // White queenside
                    self.squares[3] = EMPTY;
                    self.squares[0] = WHITE_ROOK;
                }
                62 => { // Black kingside
                    self.squares[61] = EMPTY;
                    self.squares[63] = BLACK_ROOK;
                }
                58 => { // Black queenside
                    self.squares[59] = EMPTY;
                    self.squares[56] = BLACK_ROOK;
                }
                _ => {}
            }
        }

        // Restore game state
        self.castling_rights = undo.castling_rights;
        self.en_passant_square = undo.en_passant_square;
        self.halfmove_clock = undo.halfmove_clock;

        // Update fullmove number
        if !self.white_to_move {
            self.fullmove_number -= 1;
        }

        // Remove last position from history
        self.position_history.pop();
    }

    /// Find the king's square for the specified color
    pub fn find_king(&self, white: bool) -> Option<usize> {
        let king = if white { WHITE_KING } else { BLACK_KING };
        for sq in 0..64 {
            if self.squares[sq] == king {
                return Some(sq);
            }
        }
        None
    }

    /// Count how many times the current position has occurred
    pub fn repetition_count(&self) -> usize {
        if self.position_history.is_empty() {
            return 1;
        }
        let current_hash = *self.position_history.last().unwrap();
        self.position_history.iter().filter(|&&h| h == current_hash).count()
    }

    /// Check if current position has occurred 3 times (draw by repetition)
    pub fn is_repetition(&self) -> bool {
        if self.position_history.len() < 5 {
            return false;
        }
        self.repetition_count() >= 3
    }

    /// Check if 50-move rule applies (draw)
    pub fn is_fifty_moves(&self) -> bool {
        self.halfmove_clock >= 100
    }

    /// Check for insufficient material to checkmate
    pub fn has_insufficient_material(&self) -> bool {
        let mut pieces: Vec<(u8, u8, usize)> = Vec::new();
        
        for sq in 0..64 {
            let piece = self.squares[sq];
            if piece != EMPTY {
                pieces.push((get_piece_type(piece), get_piece_color(piece), sq));
            }
        }

        // Only kings left
        if pieces.len() == 2 {
            return true;
        }

        // King and minor piece vs King
        if pieces.len() == 3 {
            for (ptype, _, _) in &pieces {
                if *ptype == KNIGHT || *ptype == BISHOP {
                    return true;
                }
            }
        }

        // King + Bishop vs King + Bishop (same color squares)
        if pieces.len() == 4 {
            let bishops: Vec<(usize, u8)> = pieces.iter()
                .filter(|(pt, _, _)| *pt == BISHOP)
                .map(|(_, c, sq)| (*sq, *c))
                .collect();
            
            if bishops.len() == 2 {
                let (sq1, c1) = bishops[0];
                let (sq2, c2) = bishops[1];
                let sq1_color = (sq1 / 8 + sq1 % 8) % 2;
                let sq2_color = (sq2 / 8 + sq2 % 8) % 2;
                if sq1_color == sq2_color && c1 != c2 {
                    return true;
                }
            }
        }

        false
    }

    /// Create a copy of the board
    pub fn copy(&self) -> Self {
        self.clone()
    }

    /// Display the board as a string
    pub fn display(&self) -> String {
        let mut lines = Vec::new();
        lines.push("  +---+---+---+---+---+---+---+---+".to_string());
        
        for rank in (0..8).rev() {
            let mut row = format!("{} |", rank + 1);
            for file in 0..8 {
                let piece = self.squares[rank * 8 + file];
                if piece == EMPTY {
                    row.push_str("   |");
                } else if let Some(c) = piece_to_fen(piece) {
                    row.push_str(&format!(" {} |", c));
                } else {
                    row.push_str(" ? |");
                }
            }
            lines.push(row);
            lines.push("  +---+---+---+---+---+---+---+---+".to_string());
        }
        lines.push("    a   b   c   d   e   f   g   h".to_string());
        
        lines.join("\n")
    }
}

impl Default for Board {
    fn default() -> Self {
        Board::new()
    }
}

impl std::fmt::Display for Board {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.display())
    }
}
